### ISSUE TITLE
[API Proposal]: TryUpdateOrRemove for the ConcurrentDictionary<K,V> collection

### ISSUE THREAD

**Issue by theodorzoulias** [dotnet/runtime#68742]
### Background and motivation

The [`ConcurrentDictionary<K,V>`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2) class offers the powerful method [`AddOrUpdate`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.addorupdate), which is an efficient and convenient way of adding or updating entries in the collection using user-supplied functions. There is no such method though for either updating or removing a key/value pair from the collection based on the key's existing value, forcing the developers to write relatively complex (spinning) methods to accomplish these tasks ([example](https://stackoverflow.com/questions/12240219/is-there-a-way-to-use-concurrentdictionary-tryupdate-with-a-lambda-expression "Is there a way to use ConcurrentDictionary.TryUpdate with a lambda expression?")). Suggestions have been made to add such APIs in the class, in the form of `TryUpdate` and `TryRemove` methods overloads, but for various reasons these suggestions became inactive and were eventually closed:

- [Suggested API Addition: ConcurrentDictionary.TryUpdate(TKey key, Func<TKey, TValue, TValue> updateFactory)](https://github.com/dotnet/runtime/issues/15539)
- [Add RemoveIfValue to ConcurrentDictionary](https://github.com/dotnet/runtime/issues/23909#issuecomment-531265755)
- [Add method RemoveOrUpdate to ConcurrentDictionary](https://github.com/dotnet/runtime/issues/23644) (still open)

This proposal is about a new `TryUpdateOrRemove` method that covers both of these functionalities, and the combination of them. It allows to update or remove a key/value pair with a single call, and a single hashcode calculation of the key. It is a slightly cumbersome API because it requires that the return value of the user-supplied lambda is a value tuple, and it also requires the introduction of a new public `enum` in the `System.Collections.Concurrent` namespace.

### API Proposal

```C#
namespace System.Collections.Concurrent
{
    /// <summary>
    /// Specifies the requested outcome of a TryUpdateOrRemove operation
    /// on a concurrent collection.
    /// </summary>
    public enum UpdateRemoveResult { NoAction, Update, Remove }

    public class ConcurrentDictionary<TKey, TValue>
    {
        /// <summary>
        /// Uses the specified function and argument to update or remove a key/value pair
        /// in the dictionary.
        /// </summary>
        /// <typeparam name="TArg">The type of an argument to pass into updateAction.</typeparam>
        /// <param name="key">The key whose value should be updated or removed.</param>
        /// <param name="updateAction">The function used to generate a new value for
        /// an existing key based on the key's existing value, or to request that the key/value
        /// pair should be removed.</param>
        /// <param name="updateActionArgument">An argument to pass into updateAction.</param>
        /// <returns>true if the object was updated or removed successfully; otherwise,
        /// false.</returns>
        public bool TryUpdateOrRemove<TArg>(
            TKey key,
            Func<TKey, TValue, TArg, (UpdateRemoveResult, TValue)> updateAction,
            TArg updateActionArgument);

        public bool TryUpdateOrRemove(
            TKey key,
            Func<TKey, TValue, (UpdateRemoveResult, TValue)> updateAction);
    }
}
```

### API Usage

Example of using this API for storing client-token information, with multiple tokens per client, and with removing inactive clients (clients with zero tokens) from the collection:

```C#
public class ClientsTokens
{
    private ConcurrentDictionary<int, ImmutableHashSet<Guid>> _activeTokens = new();

    public void AddClientToken(int clientID, Guid token)
    {
        _activeTokens.AddOrUpdate(clientID,
            (_, token) => ImmutableHashSet.Create<Guid>().Add(token),
            (_, existing, token) => existing.Add(token),
            token);
    }

    public bool RemoveClientToken(int clientID, Guid token)
    {
        return _activeTokens.TryUpdateOrRemove(clientID, (_, existing, token) =>
        {
            ImmutableHashSet<Guid> updated = existing.Remove(token);
            if (updated == existing) return (UpdateRemoveResult.NoAction, default);
            if (updated.IsEmpty) return (UpdateRemoveResult.Remove, default);
            return (UpdateRemoveResult.Update, updated);
        }, token);
    }
}
```

### Alternative Designs

Instead of the result of the function being a `(UpdateRemoveResult, TValue)`, a single `TValue` could be returned instead, and the requested action could be inferred by comparing this value with the existing value, and with the default value of the type:

1. In case the return value is equal to the existing value, take no action.
2. In case the return value equals to `default(TValue)`, remove the key/value pair.
3. In any other case, update the value of the key with the new value.

Although this design simplifies the use of the API in common scenarios, it complicates the case where the `default(TValue)` is a valid value that should be retained. It also adds an implicit dependency on the default comparer of the `TValue` type.

### Risks

None that I am aware of.

**Comment by stephentoub** [comment]
> Risks. None that I am aware of.

The risk is that it's a complicated API that's trivial to misuse.  Most use cases I've seen of folks interested in something like this aren't using an immutable collection as the value, and are very likely to get incorrect or corrupted results, given the delegate will not be invoked while holding a lock (consider what the code would look like if the value were a `HashSet<>` instead of an `ImmutableHashSet<>`).

Your example can also be written in almost as little code without the helper, and without the expense of additional delegate invocations (and while there will be multiple GetHashCode calls, your key is an Int32 for which the GetHashCode will evaporate):
```C#
public bool RemoveClientToken(int clientID, Guid token)
{
    while (_activeTokens.TryGetValue(clientID, out ImmutableHashSet<Guid> tokens))
    {
        ImmutableHashSet<Guid> removed = tokens.Remove(token);
        if (removed == tokens)
        {
            break;
        }
        else if (removed.IsEmpty)
        {
            if (_activeTokens.TryRemove(KeyValuePair.Create(clientId, tokens))
                break;
        }
        else if (_activeTokens.TryUpdate(clientId, removed, tokens))
        {
            break;
        }
    }
}
```

**Comment by theodorzoulias** [comment]
@stephentoub it's true that this API can be misused, and the `AddOrUpdate` can be misused as well. I see it all the time in StackOverflow ([very recent example](https://stackoverflow.com/questions/72060540/c-sharp-concurrent-dictionary-best-way-to-remove-collection-type-values "Concurrent dictionary best way to remove Collection type values")). If the `AddOrUpdate` method didn't exist, folks would still find other ways to misuse the `ConcurrentDictionary<K,V>`. I think it's inevitable.

The problem is that the `ConcurrentDictionary<K,V>` is routinely suggested as the collection of choice when someone wants a thread-safe `Dictionary<K,V>`. This is unlikely to be a good suggestion most of the time. The correct suggestion IMHO is a `Dictionary<K,V>`+`lock`, because it's easy to get it right if you are disciplined and know the basic principles. The `ConcurrentDictionary<K,V>` is an excellent choice with eyes closed when the `TValue` is immutable, otherwise it's very tricky to use, or just unsuitable. And most people who are following that advice are not overly familiar with the concepts of concurrency and immutability, they try to use it like a normal `Dictionary<K,V>`, and the probability of misuse approaches 100%.

I agree that the code required to achieve the `TryUpdateOrRemove` functionality is not that much, and that the cost of hashing the keys is trivial in most scenarios. But I still think that this API would be appreciated by a significant number of advanced .NET users who know what they are doing, and who fill happiness when their code is as optimized as possible!

**Comment by theodorzoulias** [comment]
@sab39 this looks like a valid alternative design, that covers a wider range of scenarios than my proposal. The scenarios where the `Modify` method could be utilized to its full potential should be quite rare though. Regarding the name, `TryModify` might be more consistent with the existing APIs than `Modify`.

Before posting this suggestion I experimented myself with a dedicated `delegate` for the `TryUpdateOrRemove` method, but I discovered that delegates with `ref` or `out` parameters have usability issues: the types of the arguments must be specified explicitly because [the compiler cannot infer them](https://stackoverflow.com/questions/4045792/how-to-pass-out-parameter-into-lambda-expression/11865853#11865853 "How to pass 'out' parameter into lambda expression"). So instead of this:

```C#
return _activeTokens.TryUpdateOrRemove(clientID, (_, existing, token) =>
```

...the consumer of the API will have to write something like this:

```C#
return _activeTokens.TryUpdateOrRemove(clientID,
    (int _, ImmutableHashSet<Guid> existing, Guid token, out ImmutableHashSet<Guid> updated) =>
```

...which is super awkward. It's tedious to write, read and maintain.
